servant/src/Servant/Types/SourceT.hs

```diff
@@ -305,13 +306,28 @@ fromAction stop action = SourceT ($ fromActionStep stop action)
 {-# INLINE fromAction #-}
 
 fromActionStep :: Functor m => (a -> Bool) -> m a -> StepT m a
-fromActionStep stop action = loop where
-    loop = Effect $ fmap step action
-    step x
-        | stop x    = Stop
-        | otherwise = Yield x loop
+fromActionStep stop action = fromConsumableActionStep action' () where
```

Comment:
I'd probably leave the definition we had before, as it didn't allocate a bunch of pairs with `()`s in them, which might start mattering when building up large streams.